File: 01-java-helloworld/src/main/java/demo/EmulatedNetworkManager.java
package demo;

import org.protelis.lang.datatype.DeviceUID;
import org.protelis.vm.CodePath;
import org.protelis.vm.NetworkManager;

import java.util.*;

public class EmulatedNetworkManager implements NetworkManager {

    private Map<CodePath, Object> toBeSent;
    private Map<DeviceUID, Map<CodePath, Object>> messages;
    private final DeviceUID uid;

    public EmulatedNetworkManager(int uid) {
        this.uid = new MyDeviceUIDImpl(uid);
        messages = new HashMap<>();
    }

    private void receiveMessage(DeviceUID src, Map<CodePath, Object> msg) {
        messages.put(src, msg);
    }

    public void sendMessages(Set<Device> neighbors) {
        neighbors.forEach(d -> d.getNetworkManager().receiveMessage(uid, toBeSent));

    }

    @Override
    public Map<DeviceUID, Map<CodePath, Object>> getNeighborState() {
        Map<DeviceUID, Map<CodePath, Object>> t = messages;
        messages = new HashMap<>();
        return t;
    }

    @Override
    public void shareState(Map<CodePath, Object> toSend) {
        toBeSent = toSend;
    }
}

File: 02-kotlin-helloworld/src/main/kotlin/EmulatedNetworkManager.kt
package demo

import org.protelis.lang.datatype.DeviceUID
import org.protelis.vm.CodePath
import org.protelis.vm.NetworkManager

class EmulatedNetworkManager(private val uid : DeviceUID) : NetworkManager {
    private var toBeSent: Map<CodePath, Any> = emptyMap()
    private var messages: Map<DeviceUID, Map<CodePath, Any>> = emptyMap()

    private fun receiveMessage(src: DeviceUID, msg: Map<CodePath, Any>) {
        messages += Pair(src, msg)
    }

    fun sendMessages(neighbors: Set<Device>) {
        if (toBeSent.isNotEmpty()) {
            neighbors.forEach { it.netmgr.receiveMessage(uid, toBeSent) }
        }
    }

    override fun getNeighborState() : Map<DeviceUID, Map<CodePath, Any>> =
        messages.apply { messages = emptyMap() }


    override fun shareState(toSend: Map<CodePath, Any>) { toBeSent = toSend }

}